bpf/lib/nat.h

```diff
@@ -1064,10 +1061,85 @@ snat_v4_nat(struct __ctx_buff *ctx, const struct ipv4_nat_target *target, __s8 *
 	if (ret < 0)
 		return ret;
 
-rewrite_egress:
 	return snat_v4_rewrite_egress(ctx, &tuple, state, off, ipv4_has_l4_header(ip4));
 }
 
+static __always_inline __maybe_unused int
+snat_v4_rev_nat_handle_icmp_frag_needed(struct __ctx_buff *ctx, __u64 off)
+{
+	struct ipv4_ct_tuple tuple = {};
+	struct ipv4_nat_entry *state;
+	struct iphdr iphdr;
+	__be16 identifier;
+	__u8 type;
+	__u32 icmpoff = off + sizeof(struct icmphdr);
+	int ret;
+
+	/* According to the RFC 5508, any networking equipment that is
+	 * responding with an ICMP Error packet should embed the original
+	 * packet in its response.
+	 */
+
+	if (ctx_load_bytes(ctx, icmpoff, &iphdr,
+			   sizeof(iphdr)) < 0)
+		return DROP_INVALID;
+
+	/* From the embedded IP headers we should be able to determine
+	 * corresponding protocol, IP src/dst of the packet sent to resolve the
+	 * NAT session.
+	 */
+	tuple.nexthdr = iphdr.protocol;
+	tuple.saddr = iphdr.daddr;
+	tuple.daddr = iphdr.saddr;
+	tuple.flags = NAT_DIR_INGRESS;
+
+	icmpoff += ipv4_hdrlen(&iphdr);
+	switch (tuple.nexthdr) {
+	case IPPROTO_TCP:
+	case IPPROTO_UDP:
+#ifdef ENABLE_SCTP
+	case IPPROTO_SCTP:
+#endif  /* ENABLE_SCTP */
+		/* No reasons to handle IP fragmentation for this case as it is
+		 * expected that DF isn't set for this particular context.
+		 */
+		if (l4_load_ports(ctx, icmpoff, &tuple.dport) < 0)
+			return DROP_INVALID;
+		break;
+	case IPPROTO_ICMP:
+		/* No reasons to see a packet different than ICMP_ECHO. */
+		if (ctx_load_bytes(ctx, icmpoff, &type, sizeof(type)) < 0 ||
+		    type != ICMP_ECHO)
+			return DROP_INVALID;
+		if (ctx_load_bytes(ctx, icmpoff +
+				   offsetof(struct icmphdr, un.echo.id),
+				   &identifier, sizeof(identifier)) < 0)
+			return DROP_INVALID;
+		tuple.sport = 0;
+		tuple.dport = identifier;
+		break;
+	default:
+		return DROP_UNKNOWN_L4;
+	}
+	state = snat_v4_lookup(&tuple);
+	if (!state)
+		return NAT_PUNT_TO_STACK;
+
+	/* We found SNAT entry to rev-NAT embedded packet. The source addr
+	 * should point to endpoint that initiated the packet, as-well if
+	 * dest port had been NATed.
+	 */
+	ret = snat_v4_icmp_rewrite_ingress_embedded(ctx, &tuple, state,
+						    off, icmpoff);
+	if (IS_ERR(ret))
+		return ret;
+
+	/* Switch back to the outer header. */
+	tuple.nexthdr = IPPROTO_ICMP;
+
+	return snat_v4_rewrite_ingress(ctx, &tuple, state, off);
+}
```

Comment:
I'm slightly concerned that this part (how we NAT the outer headers) will over time drift apart from what we do in the "main" path. How bad would it be to only handle the inner SNAT here, and return to the main function for handling the outer SNAT?